### DISCUSSION TITLE
Extend app to support climbing training

### DISCUSSION THREAD

**Discussion by adrianlzt** [wger-project/wger#1556] [Ideas]
Hi there!

I was looking for an app to replace the _typical_ Google Sheets used by trainers to send the workouts and found this nice app.

I want to know your opinion on how could be this app be extended to handle climbing, although I think most of the things could be useful for all users.

First thing are the exercises. I don't know if you agree in adding climbing-specific exercises to the app. I know I can create them in a on-prem install, but I'm not sure how those exercises could be shared between different instances.

I also use, for some exercises, a open number of sets, like, do as many sets possible until you feel a loss of perfomance. I'm not sure how that could be added. Maybe adding a lot of sets and the only log those made?

Another one related to exercises, when the same exercise should be done with one arm and then the other, the weight could be different. Maybe an extra option to differentiate between between sides (or even fingers!)?

Related with session logging I was thinking to add more options, like:
 - Rating of Perceived Exertion for each exercise
 - a pre/post session test of some kind (in climbing it is useful to save your max finger strength after warming up and at the end of the session)

That values should be shown in the log viewer adding some extra values calculated (% of force loss).

Those are easy to add (I have already added those to test how it works), but one a little bit more complicated is how to store global values, like the global max strength (to compare that value against each session).
Maybe those values could be added as measurements.

And talking about measurements, it could be nice to allow to create new values based on already stored values. Eg.: maybe I can create a pull-up speed measurement, and pull-up added weight and generate a third calculated measurement that returns the estimated 1RM. I was thinking in a python sandbox that could allow to create measurements writting python code, allowing all kinds of transformations using the already stored values.

So, what are your thoughts?

**Comment by rolandgeider** [comment]
Hi! This app actually started because I wanted to replace my sheets, had vacation and nothing to do 😅 

While we don't have RPE, we do have Reps In Reserve, which while very similar, I can imagine that it doesn't really match the exercises you'll be doing for climbing. As for the number of sets, while that's not possible, we're actually in the middle of making the workouts more flexible (#848), and this was something I hadn't thought about

When you finish a training session you can already save a general impression (bad/neutral/good) as well as write something in a free form text field, but you want something that has a more specific form and can later be queried etc?

At the moment global stuff is stored in the user profile, but that's not a good idea for everything. Adding some kind of 1RM calculator is also something that I was thinking about but your idea to model this like some kind of calculated measurements is definitely neat! Perhaps a simple config syntax would be enough?

**Comment by adrianlzt** [comment]
From my point of view, rir is how the trainer define the exercise and rpe
is your feedback to him.

If the sheet I'm currently using, I keep a global max taking the max of all
days, so maybe that's also possible, be able to access that kind of data.

I think it's hard to define the line between hard coding something and
giving enough freedom to customize.
I have seen many times the pattern of defining a language to get and with
with data, but it's hard to get it right.
Giving the possibility to use python will be more flexible, but also more
complex.
Another idea could be to allow SQL queries, but again, not easy of you are
not use to it.

So, do you think the idea of storing several values for the same exercise
(left/right, or finger 1,2,3...), could fit in the new workouts scheme?

Thanks for the response!!

**Comment by adrianlzt** [comment]
I was thinking that some of the improvements could be useful por the general app, but others don't.
I really don't like the idea of forking, so, maybe, add something like plugins that could be enabled to extend the functionality?

I'm not sure how this could be done. Should the model contain all posibilities and just some of them are enabled?
Or maybe some field/fields with dynamic values (like json in a string field) to give more flexibility?

And maybe an external _store_ of exercises where people could decide which groups to add if they want? I'm thinking in something simple, like having groups of exercises bundled in a file and available in a repo (like in https://github.com/wger-project/data).

**Comment by rolandgeider** [comment]
just added an issue for the dynamic measurements, #1560

But yes, how to present this to the user is going to be a challenge. Or if we want to go that route at all, maybe we could register some complex filters as simple python modules which the user can select but for very simple calculations there could be a UI?

At the moment we have "variations" for an exercise which are basically exercises that are similar (such as the different types of squat, curls, etc.), but those other exercises need to be different enough to get their own entry in the db. Perhaps just adding using the comments for the sets could be enough here? You'd have some exercise with a free text comment and you'd know what to do, but would not be able to run any analysis on that (plus the comments are currently not shown practically anywhere, we'd have to change that)
